src/role/participant.ts

```diff
@@ -0,0 +1,186 @@
+import { Client, Guild, Role } from "eris";
+import { MiscInternalError } from "../util/errors";
+import { getLogger } from "../util/logger";
+
+const logger = getLogger("role:participant");
+
+type ParticipantRole = {
+	id: string;
+	server: Guild;
+};
+
+type Tournament = {
+	id: string;
+	server: string;
+};
+
+/**
+ * Creates the server role that identifies confirmed status in a given tournament.
+ */
+export class ParticipantRoleProvider {
+	// Storing a reference to the Eris cached server might not be necessary if parameterised
+	protected roleCache: Record<string, ParticipantRole> = {};
+
+	// Using American spelling for Eris consistency
+	constructor(protected readonly bot: Client, readonly color?: number) {}
+
+	/**
+	 * Creates the role in question on the specified server for the specified tournament.
+	 * Exceptions (missing permissions, network fault) can be thrown.
+	 *
+	 * @param server The server to create the role in
+	 * @param tournamentId Unique label
+	 * @returns The created role
+	 */
+	private async create(server: Guild, tournamentId: string): Promise<Role> {
+		const role = await server.createRole(
+			{
+				name: `MC-${tournamentId}-player`,
+				color: this.color
+			},
+			"Auto-created by Emcee."
+		);
+		logger.verbose(`Role ${role.name} (${role.id}) created in ${server.name} (${server.id}).`);
+		return role;
+	}
+
+	/**
+	 * Retrieve the participant role and cache it, or create it if it does not exist.
+	 * Exceptions on creation (missing permissions, network fault) can be thrown.
+	 *
+	 * @param tournament Unique tournament label and Discord guild snowflake
+	 * @returns Role snowflake and guild snowflake as cached
+	 */
+	protected async lazyGet(tournament: Tournament): Promise<ParticipantRole> {
+		if (tournament.id in this.roleCache) {
+			return this.roleCache[tournament.id];
+		}
+		const server = this.bot.guilds.get(tournament.server);
+		if (!server) {
+			// TODO: determine what bizarre scenario would result in this
+			throw new MiscInternalError(
+				`Could not find server ${tournament.server} as registered with Tournament ${tournament.server}.`
```

Comment:
i think the second parameter here should be `tournament.name`, not `tournament.server`

Reply:
This issue is extant.